### ISSUE TITLE
toBeEmpty expectation for arrays

### ISSUE THREAD

**Issue by Nysosis** [alsatian-test/alsatian#293]
It would be nice if there was an expectation for array types to state whether they are empty or not, without having to check the length property (as that way we don't know from the failure whats in the array if it's not empty), because all we have is the length.

**Comment by jamesrichford** [comment]
Great idea @Nysosis, one little addition is it might be good to extend this to objects and string i.e.

```typescript

// string
Expect("").toBeEmpty();                                   // pass
Expect("something").toBeEmpty();                  // fail
Expect("something").not.toBeEmpty();            // pass
Expect("").not.toBeEmpty();                             // fail

// object
Expect({}).toBeEmpty();                                    // pass
Expect({ some: "property" }).toBeEmpty();       // fail
Expect({ some: "property" }).not.toBeEmpty(); // pass
Expect({}).not.toBeEmpty();                              // fail
```

**Comment by Nysosis** [comment]
Good thought!

Strings will be easy enough.

Objects I feel might be more finicky, because of pretty much everything in JS being an object, and the joys of the whacky prototype inheritance? Could go down the use of `for..in`, but doesn't work in all scenarios:

https://jsfiddle.net/ocdd9fxL/1/

Unless there's a trick I'm missing somewhere, or you're happy to accept edge those edgecases

**Comment by jamesrichford** [comment]
Maybe the simplest approach for now is something like

```typescript
Object.keys(instance).length === 0 && instance.constructor === Object
```

This will give us a clearer definition of empty object as:
* an object with no keys
* and is **not** an extension (i.e. Date, Error and User Defined Types are **never** empty)

**Comment by Nysosis** [comment]
Ok, sounds good, I'll get that put in when I get a chance, also realised I
didn't update the readme with the new Matcher, so I'll get that in at the
same time.

**Comment by jamesrichford** [comment]
@Nysosis no worries about the readme I noticed but as we're reworking the documentation into more of a wiki style (see #253) this won't be too much of an issue as it won't be in the readme much longer :)

**Comment by jamesrichford** [comment]
@Nysosis as mentioned above would be nice to get your opinion on #296 as to whether you think it's an improvement on the current single readme if you wouldn't mind? :)

**Comment by jamesrichford** [comment]
closed by #294 - nice one! 👍
